Other PR's merged + code improved + flow improved

mdeweerd opened the pull request:
This merges (most) of the other PR's.

I'va also "converted" the travis workflow in a ci.yml that also compiles the script with gcc sanitize options.
Those options effectively help detect leaks, and no longer detect leaks with the code in this PR.

fabio-brugnara committed:
add -P option, to pass the script via pipe instead of command line

fabio-brugnara committed:
modified untraceable() to avoid occasional freeze with zombie

fabio-brugnara committed:
add support for python*

fabio-brugnara committed:
initialize pipe flag to 0

ergoucao committed:
Fix script file names containing spaces and other special characters.

mdeweerd committed:
ci: pre-commit & generate actions

mdeweerd committed:
Fix files with linting & formatting tools + spelling correction

mdeweerd committed:
Make script executable

mdeweerd committed:
minor: Fix spelling require->requires

mdeweerd committed:
Improve format hinting

Commit:
add install procedure for Debian and Ubuntu systems in the README

Commit:
reword the 'Ubuntu-specific' heading to make it more specific

Commit:
add code block hinting for the Ubuntu PPA install procedure

ashamedbit committed:
Fix memory leaks in write_c and eval_shell

dviererbe committed:
Remove ash dependency

The ash binary package was dropped in Debian unstable, see
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=920644

Signed-off-by: Dominik Viererbe <dominik.viererbe@canonical.com>

mdeweerd committed:
Spelling updates, regenerate man

mdeweerd committed:
Merge branch 'ci/precommit_generate' into memory-leak-fixes

mdeweerd committed:
Merge branch 'fix/linting' into integration/merge

mdeweerd committed:
Merge branch 'fix/spelling/man' into integration/merge

mdeweerd committed:
Merge branch 'pr157' into integration/merge

mdeweerd committed:
Merge branch 'pr167' into integration/merge

mdeweerd committed:
Merge branch 'pr164' into integration/merge

mdeweerd committed:
Merge branch 'doc/format-hints' into integration/merge

Commit:
ci: Github Action Generate Files

mdeweerd committed:
Add exceptions for test, and possibility to exclude cppchecks

mdeweerd committed:
Exclude readability/casting because this is c-code

mdeweerd committed:
Ignore several external files for checks

mdeweerd committed:
Cleanup INSTALL

mdeweerd committed:
Cleanup .travis.yml

mdeweerd committed:
Add ci (replace travis)

mdeweerd committed:
Fix cpplint/cppcheck

mdeweerd committed:
Do test with sanitize on

mdeweerd committed:
Do test with sanitize on

mdeweerd commented:
@suntong I think this is a good base to work with.  I am not going to be the "codeowner" however.

FYI: the effectiveness of the leak detection can currently be seen in this action: https://github.com/mdeweerd/shc/actions/runs/10342670550/job/28625892548#step:4:2148

While it runs without leak reports here:
https://github.com/mdeweerd/shc/actions/runs/10342637455/job/28625817171

mdeweerd committed:
Also ignore Makefile.am for file formatting

suntong commented:
Wow, thanks a lot for such hard work @mdeweerd, that's indeed what I was hoping for, that someone tries to pick up _all_ the _possible_ enhancements before we go for a new Debian release. Really appreciate it. 

Let's give a few days for @neurobin / @ashamedbit to review, and see what they're going to say... 

thanks again.

mdeweerd committed:
Sanitizer: stop the program when a leak is detected

mdeweerd committed:
Merge commit '52b7f6b' into integration/merge

mdeweerd committed:
Avoid warning about not using the return value of setuid

mdeweerd committed:
Upgrade pandoc workflow

Commit:
ci: Github Action Generate Files

mdeweerd committed:
Add uncrustify configuration, apply it

The uncrustify configuration was set up to match the existing
layout quite closely while adding braces and indenting where
missing

mdeweerd committed:
Merge branch 'integration/merge' of github.com:mdeweerd/shc into integration/merge

mdeweerd committed:
Update executable bit for INSTALL

mdeweerd committed:
Remove space after sizeof

mdeweerd committed:
Format code before merging

mdeweerd committed:
Merge branch 'pr153' into integration/merge

fabio-brugnara committed:
add test in ttest.sh for -P flag and python

fabio-brugnara committed:
Test pip & python

mdeweerd committed:
Ignore developer name

mdeweerd committed:
Allow skipping of tests

mdeweerd committed:
Fix memory leaks in generated program

mdeweerd committed:
Fix issue with flagging overflow

mdeweerd committed:
Use unsigned to avoid runtime signed overflow notice & sprintf->snprintf

fabio-brugnara committed:
use unsigned in random generation to avoid sanitizer warning

fabio-brugnara committed:
add braces around break

mdeweerd committed:
Merge branch 'pr153' into integration/merge

fabio-brugnara committed:
add SHC_ARGV0 to environment

fabio-brugnara committed:
added SHC_PID for safer handling of SHC_ARGV0

fabio-brugnara committed:
first partial draft of argv0 fix, only for *sh

mdeweerd committed:
Test $0 value

fabio-brugnara committed:
second draft of argv0 fix, includes python, and put argv fix under option -0 (to be seen)

mdeweerd committed:
Add several test script to pre-commit checks

mdeweerd committed:
Duplicate strings before encryption to print unmangled args

mdeweerd committed:
Add flow (actions & test runs)

mdeweerd committed:
Merge branch 'argv0' of github.com:fabio-brugnara/shc into argv0

mdeweerd committed:
Disable sanitize run for fork evaluation

mdeweerd committed:
Uncrustify before merge

mdeweerd committed:
Uncrustify before merge

mdeweerd committed:
Merge branch 'argv0_with_pre' into argv0_test

mdeweerd committed:
Set -0 option in tests

mdeweerd committed:
Enable -0 option in tests

mdeweerd committed:
Enable -0 option in tests

fabio-brugnara committed:
fix typo in python3 DB

fabio-brugnara committed:
add perl support (with -P forced)

mdeweerd committed:
Add shell name and option to test temporary path

mdeweerd committed:
Add shell name and option to test temporary path

mdeweerd committed:
Add timeout value to ci run

mdeweerd committed:
Add timeout value to ci run

mdeweerd committed:
Fix timeout syntax & rename job

mdeweerd committed:
Extend tests with arguments with spaces and quotes, and fix tcsh and csh for that

fabio-brugnara committed:
fix sourcing of binary for *csh

mdeweerd committed:
Solve $0 for zsh when using pipe

mdeweerd committed:
Adjust test to not require forged $0 for rc and the pipe option

fabio-brugnara committed:
fix minor oversight in shell list

mdeweerd committed:
Merge branch 'argv0' into argv0_with_pre

mdeweerd committed:
Merge branch 'argv0_with_pre' into argv0_test

mdeweerd committed:
Remove temporary output for debug from ttest.sh

Commit:
ci: Github Action Generate Files

mdeweerd committed:
Add dependabot configuration

mdeweerd committed:
Fix executable flag on generated files

dependabot[bot] committed:
Bump dorny/paths-filter from 2 to 3

Bumps [dorny/paths-filter](https://github.com/dorny/paths-filter) from 2 to 3.
- [Release notes](https://github.com/dorny/paths-filter/releases)
- [Changelog](https://github.com/dorny/paths-filter/blob/master/CHANGELOG.md)
- [Commits](https://github.com/dorny/paths-filter/compare/v2...v3)

---
updated-dependencies:
- dependency-name: dorny/paths-filter
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

mdeweerd committed:
Merge pull request #1 from mdeweerd/dependabot/github_actions/dorny/paths-filter-3

Bump dorny/paths-filter from 2 to 3

fabio-brugnara committed:
make forging of argv0 opt-out with -p, force -P for *csh

fabio-brugnara committed:
Change meaning of options -P -p. Added set of __file__ to python preamble

mdeweerd committed:
Update Changelog

mdeweerd committed:
Merge branch 'argv0' into argv0_with_pre

mdeweerd committed:
Merge branch 'argv0_with_pre' into integration/merge

mdeweerd committed:
Cleanup format of generated script

mdeweerd committed:
Extend tests to more options

mdeweerd committed:
Update documentation

Commit:
ci: Github Action Generate Files

mdeweerd committed:
Avoid warning about unused return value of system() + doc

mdeweerd committed:
Ignore some file to avoid notices in ci.

mdeweerd committed:
Make changes to move the repository

mdeweerd committed:
Update man for move

invalid-email-address committed:
ci: Github Action Generate Files

mdeweerd committed:
Fix for Cygwin

mdeweerd committed:
Ignore _SC_ARG_MAX check on cygwin

mdeweerd committed:
WIP release action

mdeweerd committed:
Checkout head for modification

mdeweerd committed:
Test tagging differently

mdeweerd committed:
log github vars

mdeweerd committed:
Test ref_name

mdeweerd committed:
Correct method to get branch

mdeweerd committed:
Correct line join

mdeweerd committed:
:wrench: Update dependabot configuration

mdeweerd committed:
:wrench: Update pre-commit hooks versions

Update versions of pre-commit hooks to their latest stable releases:
- mdformat: 0.7.17 -> 1.0.0
- pre-commit-hooks: v4.6.0 -> v6.0.0
- beautysh: v6.2.1 -> v6.4.3
- codespell: v2.3.0 -> v2.4.2
- shellcheck-py: v0.10.0.1 -> v0.11.0.1

This update ensures we're using the latest features and security patches for our pre-commit hooks.

mdeweerd committed:
remove dependabot nix = flake.nix, not used

dependabot[bot] committed:
Bump cpplint from 1.6.1 to 2.0.2 in the infrastructure group

Bumps the infrastructure group with 1 update: cpplint.


Updates `cpplint` from 1.6.1 to 2.0.2
Bump the infrastructure group with 6 updates

Bumps the infrastructure group with 6 updates:

| Package | From | To |
| --- | --- | --- |
| [actions/checkout](https://github.com/actions/checkout) | `4` | `7` |
| [actions/upload-artifact](https://github.com/actions/upload-artifact) | `4` | `7` |
| [dorny/paths-filter](https://github.com/dorny/paths-filter) | `3` | `4` |
| [actions/setup-python](https://github.com/actions/setup-python) | `5` | `6` |
| [actions/cache](https://github.com/actions/cache) | `4` | `6` |
| [mdeweerd/logToCheckStyle](https://github.com/mdeweerd/logtocheckstyle) | `2024.3.5` | `2026.7.1` |


Updates `actions/checkout` from 4 to 7
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v4...v7)

Updates `actions/upload-artifact` from 4 to 7
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v4...v7)

Updates `dorny/paths-filter` from 3 to 4
- [Release notes](https://github.com/dorny/paths-filter/releases)
- [Changelog](https://github.com/dorny/paths-filter/blob/master/CHANGELOG.md)
- [Commits](https://github.com/dorny/paths-filter/compare/v3...v4)

Updates `actions/setup-python` from 5 to 6
- [Release notes](https://github.com/actions/setup-python/releases)
- [Commits](https://github.com/actions/setup-python/compare/v5...v6)

Updates `actions/cache` from 4 to 6
- [Release notes](https://github.com/actions/cache/releases)
- [Changelog](https://github.com/actions/cache/blob/main/RELEASES.md)
- [Commits](https://github.com/actions/cache/compare/v4...v6)

Updates `mdeweerd/logToCheckStyle` from 2024.3.5 to 2026.7.1
- [Commits](https://github.com/mdeweerd/logtocheckstyle/compare/v2024.3.5...v2026.7.1)

---
updated-dependencies:
- dependency-name: cpplint
  dependency-version: 2.0.2
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: infrastructure
- dependency-name: actions/checkout
  dependency-version: '7'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: infrastructure
- dependency-name: actions/upload-artifact
  dependency-version: '7'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: infrastructure
- dependency-name: dorny/paths-filter
  dependency-version: '4'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: infrastructure
- dependency-name: actions/setup-python
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: infrastructure
- dependency-name: actions/cache
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: infrastructure
- dependency-name: mdeweerd/logToCheckStyle
  dependency-version: 2026.7.1
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: infrastructure
...

Signed-off-by: dependabot[bot] <support@github.com>

mdeweerd committed:
Merge pull request #3 from mdeweerd/dependabot/infrastructure-a016b7dcfa

Bump the "infrastructure" group with 2 updates across multiple ecosystems

mdeweerd committed:
:wrench: Update codespell ignore-words-list

Added 'indx' to the codespell ignore-words-list to prevent false positives for the word 'index'

mdeweerd committed:
Fix cache issue on cppcheck

mdeweerd committed:
:recycle: Refactor shc.c to improve code quality

- Renamed variable 'opts' to 'arg_opts' for clarity
- Added const qualifiers to function parameters to indicate immutability
- Added comments to suppress cppcheck warnings
- Improved error messages by using more descriptive variable names

mdeweerd committed:
:art: Fix typo in uncrustify.cfg

Corrected the typo in the comment section of uncrustify.cfg from 'Newline forceing' to 'Newline forcing' to maintain consistent spelling.

mdeweerd committed:
Update to new pre-commit-hooks
